File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here

  let newArr = -1

  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length - 1; j++) {
      if (arr[i] === arr[j]) {
        newArr = arr[i]
      }
    }
  }

  return newArr
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file

//// Create variable = -1
//// Iterate i (starts from index 0) and j (index 1)
//// If matched, var = i


// THIRD ATTEMPT

// newArr = -1;

// arr.find((currentValue, currentIndex) =>
//   arr.indexOf(currentValue) !== currentIndex ? (newArr = currentValue) : null
// );

// return newArr;



//// SECOND ATTEMPT
// I want to return first duplicate value in array
// First, array.find => if arr is not unique by comparing each value's index
// Then, return that value

// // Using .find ///

// let newArr = -1;
// arr.find((item, index) =>
//   arr.indexOf(item) !== index ? (newArr = item) : null
// );

// return newArr

//// Using for loop ///

// let newArr = -1;

// for (let i = 0; i < arr.length - 1; i++) {
//   for (let j = i + 1; j < arr.length; j++) {
//     if (arr[i] === arr[j]) {
//       newArr = arr[i]
//     }
//   }
// }

// return newArr

//// FIRST ATTEMPT
//// I want to iterate each number
//// Then, compare iterated number with variable =[]
//// If not, add to []

File: ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  # type your code in here

  newArr = Set.new
  i = 0


  arr.each do |value|
    if newArr.include?(value) 
      return value
    else
      newArr << value
    end
  end

  -1

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file



  #### SECOND ATTEMPT
  #### Create empty set -->  #<Set: {}>  will look like this after addition
  ####  #<Set: {1, 3, 4} >
  #### Then iterate each value 
  #### Return if set includes iterated value, if not, add value to set

  # uniques = Set.new

  # arr.each do |value|
  #   if uniques.include?(value)
  #    return value 
  #   else
  #    uniques.add(value)
  #   end
  # end

  # -1



#### FIRST ATTEMPT

## Create variable = -1
## I want to iterate number i and j and find matching

  # newArr = []

  # arr.each do |e|
  #   return e if newArr[e] ## <- e is index number
  #   newArr[e] = true
  # end

  # -1
